incubator/yarn-plugin-external-workspaces/src/utilities.ts

```diff
@@ -0,0 +1,62 @@
+import { structUtils, type Descriptor, type Locator } from "@yarnpkg/core";
+import path from "node:path";
+
+const externalProtocol = "external:";
+const remoteProtocol = "fallback:";
+
+export function getProtocol() {
+  return externalProtocol;
+}
+
+export function getRemoteProtocol() {
+  return remoteProtocol;
+}
+
+export function toPortableRelativePath(from: string, to: string): string {
+  return path.posix.normalize(path.relative(from, to));
+}
+
+export type DescriptorRangeParts = {
+  // the protocol, e.g. 'external:', empty string if no protocol
+  protocol: string;
+
+  // the version, e.g. '^1.2.3' or '*', should exist
+  version: string;
+};
+
+/**
+ * @param range the range to decode, either "protocol:version" or "version"
+ * @returns the version and protocol or an empty string for protocol if it isn't present
+ */
+export function decodeDescriptorRange(range: string): DescriptorRangeParts {
+  const protocolEnd = range.indexOf(":");
+  if (protocolEnd !== -1) {
```

Comment:
Should this be greater than zero to handle strings that just have `:` prefixed for whatever reason?
```suggestion
  if (protocolEnd > 0) {
```

Reply:
The !== -1 is correct actually. In the case of ":^3.2.1" it will end up with a procol of an empty string and version of "^3.2.1" which is correct. If it was > 0 you would end up with a version of ":^3.2.1" which would fail semver validation.